### ISSUE TITLE
Mocking libraries: NSubstitute and RhinoMocks

### ISSUE THREAD

**Issue by ChrisHurley** [red-gate/internal-tech-radar#8]
I'd propose that NSubstitute is a viable alternative to Moq (they work similarly in practice but NSubstitute has less boilerplate/lambdas), whereas RhinoMocks does not seem to be maintained any longer and seems to have few reasons to recommend it over other libraries.

**Comment by angellaa** [comment]
I had a quick look at NSubstitute and it looks great and very intuitive. From the documentation I can't think of something that is not covered. In particular I like to see they also implement auto and recursive mocks that is something moq support but few people knows about (using the moq fluent syntax). I never used NSubstitute but it is certainly something to experiment with.

**Comment by angellaa** [comment]
Definitely agree about RhinoMocks. It should be in the no more of this section RhinoMocks

**Comment by fffej** [comment]
I think RhinoMocks is used extensively in CORE? (CC @musmuris @matburton). Would you recommend it?

**Comment by matburton** [comment]
It's fine @fffej in that it's got the features we need and it's stable.
It's very similar to `Moq` and we only settled on it because we had more `RhinoMocks` test code than `Moq` code when we decided to pick one :wink:
I _do_ think `Moq` is very slightly more verbose and restrictive, but that _isn't_ what drive our decision.
Then again `RhinoMocks` can be more confusing at times (because of extra features).

**tl;dr**
![image](https://cloud.githubusercontent.com/assets/2635288/16691651/5c2c96d4-4525-11e6-8f0c-ea64b322d112.png)

**Comment by alastairs** [comment]
I use NSubstitute for all new projects. It's a really nice library for simple mocking tasks. It is, I think, less fully-featured than Moq, and when you step a little too far from its core functionality it gets _very_ clunky.

As it happens, I did a comparison of precisely these three frameworks for @matburton a few weeks back. https://gist.github.com/alastairs/c6c2ffb1bb0e67a179f4562a59ed168d/revisions

In short, I would recommend:

1. use NSubstitute for new projects, and keep to simple use of mocks (see JB Rainsberger's [Collaboration and Contract tests](https://www.youtube.com/watch?v=VDfX44fZoMc), for example)
2. use Moq for existing projects where complex stub and mock setup is already in place.

**Comment by adrianbanks** [comment]
I've also used NSubstitute in the past and it is now my preferred choice (having using all of Rhino, Moq, NSubstitute, and FakeItEasy), mainly because of its concise syntax. Whenever I've hit the problem @alastairs described (complex setup) where it gets clunky, I've taken it as a code smell in my tests and refactored the code being tested to make it more testable, eliminating the need for the complex setup.

At my last place, we had lots (> 20,000) tests that used Rhino. We decided to go with NSubstitute instead, and gradually moved over to it. We did this by having both libraries referenced, writing new fixtures in NSubstitute, and keeping old tests in Rhino except if significant change was happening to them.

Definitely agree no more Rhino for new stuff - it's dead as far as its original creator is concerned (although it has passed to a new owner), and also has an unnecessary too complex syntax compared to more modern libraries that do the same thing.

**Comment by fffej** [comment]
I've tried to represent this in the tech radar in this PR https://github.com/red-gate/internal-tech-radar/pull/26.

If there's agreement, then feel free to merge into master.
